legend-engine-language-pure-dsl-persistence/src/main/java/org/finos/legend/engine/language/pure/dsl/persistence/grammar/from/PersistenceParseTreeWalker.java

```diff
@@ -0,0 +1,695 @@
+    {
+        this.section.imports = ListIterate.collect(ctx.imports().importStatement(), importCtx -> PureGrammarParserUtility.fromPath(importCtx.packagePath().identifier()));
+        ctx.persistencePipe().stream().map(this::visitPersistencePipe).peek(e -> this.section.elements.add(e.getPath())).forEach(this.elementConsumer);
+    }
+
+    private PersistencePipe visitPersistencePipe(PersistenceParserGrammar.PersistencePipeContext ctx)
+    {
+        PersistencePipe persistencePipe = new PersistencePipe();
+        persistencePipe.name = PureGrammarParserUtility.fromIdentifier(ctx.qualifiedName().identifier());
+        persistencePipe._package = ctx.qualifiedName().packagePath() == null ? "" : PureGrammarParserUtility.fromPath(ctx.qualifiedName().packagePath().identifier());
+        persistencePipe.sourceInformation = walkerSourceInformation.getSourceInformation(ctx);
+
+        // documentation
+        PersistenceParserGrammar.DocumentationContext documentationContext = PureGrammarParserUtility.validateAndExtractRequiredField(ctx.documentation(), "doc", persistencePipe.sourceInformation);
+        persistencePipe.documentation = PureGrammarParserUtility.fromGrammarString(documentationContext.STRING().getText(), true);
+
+        // owners (optional)
+        PersistenceParserGrammar.OwnersContext ownersContext = PureGrammarParserUtility.validateAndExtractOptionalField(ctx.owners(), "owners", persistencePipe.sourceInformation);
+        persistencePipe.owners = ownersContext != null && ownersContext.STRING() != null ? ListIterate.collect(ownersContext.STRING(), ownerCtx -> PureGrammarParserUtility.fromGrammarString(ownerCtx.getText(), true)) : Collections.emptyList();
+
+        // trigger
+        PersistenceParserGrammar.TriggerContext triggerContext = PureGrammarParserUtility.validateAndExtractRequiredField(ctx.trigger(), "trigger", persistencePipe.sourceInformation);
+        persistencePipe.trigger = visitTrigger(triggerContext);
+
+        // reader
+        PersistenceParserGrammar.ReaderContext readerContext = PureGrammarParserUtility.validateAndExtractRequiredField(ctx.reader(), "reader", persistencePipe.sourceInformation);
+        persistencePipe.reader = visitReader(readerContext);
+
+        // persister
+        PersistenceParserGrammar.PersisterContext persisterContext = PureGrammarParserUtility.validateAndExtractRequiredField(ctx.persister(), "persister", persistencePipe.sourceInformation);
+        persistencePipe.persister = visitPersister(persisterContext);
+
+        return persistencePipe;
+    }
+
+    private Trigger visitTrigger(PersistenceParserGrammar.TriggerContext ctx)
+    {
+        if (ctx.TRIGGER_OPAQUE() != null)
+        {
+            OpaqueTrigger opaqueTrigger = new OpaqueTrigger();
+            opaqueTrigger.sourceInformation = walkerSourceInformation.getSourceInformation(ctx);
+
+            return opaqueTrigger;
+        }
+        throw new UnsupportedOperationException();
```

Comment:
Rather than UOE, I suggest we throw EngineException, so that studio can know where the grammar mistake is:

```java
throw new EngineException("Unsupported trigger",  walkerSourceInformation.getSourceInformation(ctx);, EngineErrorType.PARSER);
```

Reply:
Good catch, fixing ...

Reply:
Switched to EngineException, but messages don't make it back to Studio. Extended ParserErrorListener to report alternatives for some parse exceptions.

https://github.com/davidsunglee/legend-engine/blob/8bd67f9c053ff75eb8bc1040bffe4a171c38bf3e/legend-engine-language-pure-grammar/src/main/java/org/finos/legend/engine/language/pure/grammar/from/ParserErrorListener.java#L43-L54

Reply:
Got it!  I mean more if you compile (F9) on studio, does not studio take you to text mode and highlight the errors?

Reply:
It does, but I wasn't able to get error messages specified in the TreeWalker to bubble up (only get 'Unexpected token'). Might be me!

The change above will display legal tokens:

<img width="2560" alt="Screen Shot 2022-02-17 at 5 25 00 PM" src="https://user-images.githubusercontent.com/24470/154581287-b7e259a6-e51f-4987-917a-21ea46d1d4c7.png">